akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala

```diff
@@ -0,0 +1,1577 @@
+          true
+        case BufferedCommand(u @ Update(_, readC, writeC, req), replyTo) ⇒
+          receiveUpdate(key, u.modify, readC, writeC, req, replyTo)
+          (readC == ReadLocal)
+        case BufferedCommand(Delete(_, consistency), replyTo) ⇒
+          receiveDelete(key, consistency, replyTo)
+          true
+      }
+      if (cont && remaining.nonEmpty) drain()
+    }
+
+    if (updateInProgressBuffer(key).nonEmpty) drain()
+    else updateInProgressBuffer -= key
+
+    if (updateInProgressBuffer.isEmpty) context.become(normalReceive)
+  }
+
+  def receiveWrite(key: String, envelope: DataEnvelope): Unit = {
+    write(key, envelope)
+    sender() ! WriteAck
+  }
+
+  def write(key: String, writeEnvelope: DataEnvelope): Unit =
+    getData(key) match {
+      case Some(DataEnvelope(DeletedData, _)) ⇒ // already deleted
+      case Some(envelope @ DataEnvelope(existing, _)) ⇒
+        if (existing.getClass == writeEnvelope.data.getClass || writeEnvelope.data == DeletedData) {
+          val merged = envelope.merge(pruningCleanupTombstoned(writeEnvelope)).addSeen(selfAddress)
+          setData(key, merged)
+        } else {
+          log.warning("Wrong type for writing [{}], existing type [{}], got [{}]",
+            key, existing.getClass.getName, writeEnvelope.data.getClass.getName)
+        }
+      case None ⇒
+        setData(key, pruningCleanupTombstoned(writeEnvelope).addSeen(selfAddress))
+    }
+
+  def receiveReadRepair(key: String, writeEnvelope: DataEnvelope): Unit = {
+    write(key, writeEnvelope)
+    sender() ! ReadRepairAck
+  }
+
+  def receiveGetKeys(): Unit = {
+    val keys: Set[String] = dataEntries.collect {
+      case (key, (DataEnvelope(data, _), _)) if data != DeletedData ⇒ key
+    }(collection.breakOut)
+    sender() ! GetKeysResult(keys)
+  }
+
+  def receiveDelete(key: String, consistency: WriteConsistency, replyTo: ActorRef): Unit = {
+    // don't use sender() in this method, since it is used from drainUpdateInProgressBuffer
+    getData(key) match {
+      case Some(DataEnvelope(DeletedData, _)) ⇒
+        // already deleted
+        replyTo ! DataDeleted(key)
+      case _ ⇒
+        setData(key, DeletedEnvelope)
+        if (isLocalUpdate(consistency))
+          replyTo ! DeleteSuccess(key)
+        else
+          context.actorOf(WriteAggregator.props(key, DeletedEnvelope, consistency, None, nodes, replyTo))
+    }
+  }
+
+  def setData(key: String, envelope: DataEnvelope): Unit = {
+    // notify subscribers, later
+    if (subscribers.contains(key) && !changed.contains(key)) {
+      val oldDigest = getDigest(key)
+      changed = changed.updated(key, oldDigest)
+    }
+
+    val dig = if (envelope.data == DeletedData) DeletedDigest else LazyDigest
+    dataEntries = dataEntries.updated(key, (envelope, dig))
+  }
+
+  def getDigest(key: String): Digest = {
+    dataEntries.get(key) match {
+      case Some((envelope @ DataEnvelope(_, _), LazyDigest)) ⇒
```

Comment:
this match can be simplified to
~~~scala
case Some((envelope, LazyDigest)) =>
~~~